ecommerce/api.py

```diff
@@ -310,3 +318,48 @@ def enroll_user_on_success(order):
 
     if exceptions:
         raise EcommerceEdxApiException(exceptions)
+
+
+def is_coupon_redeemable_for_run(coupon, user, course_key):
+    """
+    Returns true if the coupon is redeemable for the user for a course run.
+    """
+    if not is_coupon_redeemable(coupon, user):
+        return False
+
+    return course_key in coupon.course_keys
+
+
+def is_coupon_redeemable(coupon, user):
+    """
+    Returns true if the coupon is redeemable for the user, for any relevant course run.
+    """
+    if not Program.objects.filter(
+            programenrollment__user=user,
+            programenrollment__program__course__courserun__edx_course_key__in=coupon.course_keys,
+            live=True,
+    ).exists():
+        return False
+
+    if (
+            not coupon.is_valid or                      # coupon must be enabled and within valid date range
+            not coupon.user_has_redemptions_left(user)  # coupon must not be used up
+    ):
+        return False
+
+    if coupon.coupon_type == Coupon.DISCOUNTED_PREVIOUS_COURSE:
+        # We validate in clean() that content_object is a Course if coupon_type is DISCOUNTED_PREVIOUS_RUN
+        course = coupon.content_object
+        program = course.program
+        edx_user_data = CachedEdxUserData(user, program=program)
+
+        mmtrack = MMTrack(
+            user,
+            program,
+            edx_user_data,
+        )
+
+        # For this coupon type the user must have already purchased a course run on edX
+        return any((mmtrack.has_verified_cert(run.edx_course_key) for run in course.courserun_set.all()))
+
+    return True
```

Comment:
I'm on the fence as to whether the default behavior should be for a valid coupon rather than defaulting to `False` and returning `True` for specific cases. I can see that making this code far less elegant though.